components/brave_extension/extension/brave_extension/background/today/publishers.ts

```diff
@@ -0,0 +1,125 @@
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Events from '../../../../../common/events'
+import { URLS } from './privateCDN'
+import { getPrefs as getPublisherPrefs, addPrefsChangedListener } from './publisher-user-prefs'
+
+let memoryData: BraveToday.Publishers | undefined
+let readLock: Promise<void> | null
+const url = URLS.braveTodayPublishers
+const publishersEvents = new Events()
+const eventNameChanged = 'publishers-changed'
+const storageKey = 'todayPublishers'
+const STORAGE_SCHEMA_VERSION = 1
+
+function isValidStorageData (data: {[key: string]: any}) {
+  return (
+    data && data.todayPublishers &&
+    data.todayPublishers.storageSchemaVersion === STORAGE_SCHEMA_VERSION &&
+    data.todayPublishers.publishers
+  )
+}
+
+function setPublishersCache (publishers: BraveToday.Publishers) {
+  chrome.storage.local.set({
+    [storageKey]: {
+      storageSchemaVersion: STORAGE_SCHEMA_VERSION,
+      publishers
+    }
+  })
+}
+
+function getPublishersFromCache () {
+  return new Promise<void>(resolve => {
+    chrome.storage.local.get(storageKey, (data) => {
+      if (isValidStorageData(data)) {
+        memoryData = data[storageKey].publishers
+      }
+      resolve()
+    })
+  })
+}
+
+const getLocalDataLock: Promise<void> = getPublishersFromCache()
+
+async function convertToObject(publishers: BraveToday.Publisher[]): Promise<BraveToday.Publishers> {
+  const prefs = await getPublisherPrefs()
+  const data: BraveToday.Publishers = {}
+  for (const publisher of publishers) {
+    data[publisher.publisher_id] = {
+      ...publisher,
+      // TODO: read existing value
+      user_enabled: prefs[publisher.publisher_id]
+    }
+  }
+  return data
+}
+
+function performUpdate() {
+  // Sanity check
+  if (readLock) {
+    console.error('Asked to update feed but already waiting for another update!')
+    return
+  }
+  // Only run this once at a time, otherwise wait for the update
+  readLock = new Promise(async function (resolve, reject) {
+    try {
+      const feedResponse = await fetch(url)
+      if (feedResponse.ok) {
+        const feedContents: BraveToday.Publisher[] = await feedResponse.json()
+        console.log('Got today publishers', feedContents)
+        memoryData = await convertToObject(feedContents)
+        resolve()
+        console.log('setting today publishers data', memoryData)
+        // Notify
+        setPublishersCache(memoryData)
+        publishersEvents.dispatchEvent<BraveToday.Publishers>(eventNameChanged, memoryData)
+      } else {
+        throw new Error(`Not ok when fetching publishers. Status ${feedResponse.status} (${feedResponse.statusText})`)
+      }
+    } catch (e) {
+      console.error('Could not process publishers contents from ', url)
+      reject(e)
+    } finally {
+      readLock = null
+    }
+  })
+}
+
+export async function getOrFetchData() {
+  await getLocalDataLock
+  if (memoryData) {
+    return memoryData
+  } else {
```

Comment:
maybe a stylistic thing - so feel free to ignore 😛 But when doing a boolean check, it can sometimes be more readable to just do the second action instead of putting into an else. ex:
```
if (memoryData) {
    return memoryData
}
return await update()
```

Reply:
Thanks for spotting, probably just how it iterated in this case